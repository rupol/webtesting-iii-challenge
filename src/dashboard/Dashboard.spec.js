// Test away
import React from "react";
import * as rtl from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import Dashboard from "./Dashboard";

afterEach(rtl.cleanup);

test("<Dashboard /> snapshot", () => {
  const dashboard = rtl.render(<Dashboard />);
  expect(dashboard.asFragment()).toMatchSnapshot();
});

// shows the controls and display
test("shows the controls and display", () => {
  const dashboard = rtl.render(<Dashboard />);
  const lock = dashboard.getByText(/lock gate/i);
  const close = dashboard.getByText(/close gate/i);
  const unlocked = dashboard.getByText(/unlocked/i);
  const open = dashboard.getByText(/open/i);

  expect(lock).toBeVisible();
  expect(close).toBeVisible();
  expect(unlocked).toBeVisible();
  expect(open).toBeVisible();
});

// buttons' text changes to reflect the state the door will be in if clicked
test("Button text changes on click", () => {
  const wrapper = rtl.render(<Dashboard />);
  const gateButton = wrapper.getByTestId("gate");
  const lockButton = wrapper.getByTestId("lock");

  // click "close gate"
  rtl.act(() => {
    rtl.fireEvent.click(gateButton);
  });

  // button text changes to "open gate"
  expect(gateButton).toHaveTextContent(/open gate/i);
  expect(lockButton).toHaveTextContent(/lock gate/i);

  // click "lock gate"
  rtl.act(() => {
    rtl.fireEvent.click(lockButton);
  });

  // button text changes to "unlock gate"
  expect(lockButton).toHaveTextContent(/unlock gate/i);
  expect(gateButton).toHaveTextContent(/open gate/i);
});

// the closed toggle button is disabled if the gate is locked
// the locked toggle button is disabled if the gate is open
test("Buttons disabled functionality", () => {
  const wrapper = rtl.render(<Dashboard />);
  const gateButton = wrapper.getByTestId("gate");
  const lockButton = wrapper.getByTestId("lock");

  // default - "lock gate" is disabled
  expect(lockButton).toHaveProperty("disabled", true);
  expect(gateButton).toHaveProperty("disabled", false);

  // click "close gate"
  rtl.act(() => {
    rtl.fireEvent.click(gateButton);
  });

  // click "lock gate"
  rtl.act(() => {
    rtl.fireEvent.click(lockButton);
  });

  // after clicks - "open gate" is disabled
  expect(gateButton).toHaveProperty("disabled", true);
  expect(lockButton).toHaveProperty("disabled", false);
});
