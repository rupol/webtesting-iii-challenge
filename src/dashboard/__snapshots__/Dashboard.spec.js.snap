// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dashboard /> snapshot 1`] = `
<DocumentFragment>
  <div
    class="display panel"
  >
    <div
      class="led green-led"
    >
      Unlocked
    </div>
    <div
      class="led green-led"
    >
      Open
    </div>
  </div>
  <div
    class="controls panel"
  >
    <button
      class="toggle-btn"
      data-testid="lock"
      disabled=""
    >
      Lock Gate
    </button>
    <button
      class="toggle-btn"
      data-testid="gate"
    >
      Close Gate
    </button>
  </div>
</DocumentFragment>
`;
